Framework/script/RepoCleaner/rules/production.py

```diff
@@ -0,0 +1,226 @@
+    deletion_list: List[ObjectVersion] = []
+    update_list: List[ObjectVersion] = []
+    runs_dict: DefaultDict[str, List[ObjectVersion]] = defaultdict(list)
+
+    # Extra parameters
+    delay_first_trimming = int(extra_params.get("delay_first_trimming", 30))
+    logging.debug(f"delay_first_trimming : {delay_first_trimming}")
+    period_btw_versions_first = int(extra_params.get("period_btw_versions_first", 10))
+    logging.debug(f"period_btw_versions_first : {period_btw_versions_first}")
+    delay_final_trimming = int(extra_params.get("delay_final_trimming", 180))
+    logging.debug(f"delay_final_trimming : {delay_final_trimming}")
+    period_btw_versions_final = int(extra_params.get("period_btw_versions_final", 60))
+    logging.debug(f"period_btw_versions_final : {period_btw_versions_final}")
+
+    # Find all the runs and group the versions
+    versions = ccdb.getVersionsList(object_path)
+    logging.debug(f"Dispatching versions to runs")
+    for v in versions:
+        if "RunNumber" in v.metadata:
+            runs_dict[v.metadata['RunNumber']].append(v)
+        else:
+            runs_dict[-1].append(v)  # the ones with no run specified
+    logging.debug(f"   Number of runs : {len(runs_dict)}")
+    logging.debug(f"   Number of versions without runs : {len(runs_dict[-1])}")
+
+    # Versions without runs: spare if more recent than the delay
+    logging.debug(f"Eliminating versions without runs if older than the grace period")
+    for run_version in runs_dict[-1]:
+        if in_grace_period(run_version, delay):
+            preservation_list.append(run_version)
+        else:
+            logging.debug(f"   delete {run_version}")
+            deletion_list.append(run_version)
+            ccdb.deleteVersion(run_version)
+    del runs_dict[-1]  # remove this "run" from the list
+
+    # For each run
+    logging.debug(f"Trimming the versions with a run number")
+    for run, run_versions in runs_dict.items():
+        logging.debug(f"   Processing run {run}")
+        # TODO get the EOR if it happened, meanwhile we use `eor_dict` or compute first object time + 15 hours
+        eor = eor_dict.get(int(run), run_versions[0].validFromAsDt + timedelta(hours=15))
+        logging.debug(f"   EOR : {eor}")
+
+        # run is finished for long enough
+        if eor is not None and datetime.now() > eor + timedelta(minutes=delay_final_trimming):
+            logging.debug("      Run is over for long enough, let's do the final trimming")
+            final_trimming(ccdb, period_btw_versions_final, run_versions, preservation_list, update_list, deletion_list)
+
+        else:  # trim the versions as the run is ongoing or too fresh
+            logging.debug("      Run is too fresh or still ongoing, we do the light trimming")
+            first_trimming(ccdb, delay_first_trimming, period_btw_versions_first, run_versions,
+                           preservation_list, update_list, deletion_list)
```

Comment:
will this actually behave correctly with an empty line between if and else?

Reply:
yes, it does but you are correct that it does not respect the usual coding standard of python. Fixed.